### ISSUE TITLE
OOP Background Task doesn't run in WinUI3/WinAppSDK desktop app

### ISSUE THREAD

**Issue by JosHuybrighs** [microsoft/WindowsAppSDK#4113]
### Describe the bug

I have a beta version available of my WinUI3/WinAppSDK desktop app for a restricted set of users via Microsoft Store Package Flights. One of the users reports that none of the 4 background tasks in the app are executed. The 4 type of tasks are: a 15 minutes timer triggered task, a portable device watcher task, a session connect watcher task, and a package update task. All 4 are OOP BG tasks and run without issues on the other user's machines. The manifest file of course is correctly set up with the necessary entries:

```
<Application>
   <Extensions>
        <Extension Category="windows.backgroundTasks" EntryPoint="SyncFolder.BackgroundTasks.ScheduleBackupTask">
          <BackgroundTasks>
            <Task Type="timer"/>
          </BackgroundTasks>
        </Extension>
        <!--Register session connected background tasks-->
        <Extension Category="windows.backgroundTasks"  EntryPoint="SyncFolder.BackgroundTasks.ExecWhenSessionConnectedTask">
          <BackgroundTasks>
            <Task Type="systemEvent"/>
          </BackgroundTasks>
        </Extension>
        <!--Register portable device watcher background tasks-->
        <Extension Category="windows.backgroundTasks" EntryPoint="SyncFolder.BackgroundTasks.PortableDeviceWatcherTask">
          <BackgroundTasks>
            <Task Type="deviceUse"/>
          </BackgroundTasks>
        </Extension>
        <!--Register update background tasks-->
        <Extension Category="windows.updateTask" EntryPoint="SyncFolder.BackgroundTasks.UpdateTask"/>
...

```
And:
```
  <Extensions>
    <Extension Category="windows.activatableClass.inProcessServer">
      <InProcessServer>
        <Path>SyncFolder\WinRT.Host.dll</Path>
        <ActivatableClass ActivatableClassId="SyncFolder.BackgroundTasks.ScheduleBackupTask" ThreadingModel="both"/>
      </InProcessServer>
    </Extension>
    <Extension Category="windows.activatableClass.inProcessServer">
      <InProcessServer>
        <Path>SyncFolder\WinRT.Host.dll</Path>
        <ActivatableClass ActivatableClassId="SyncFolder.BackgroundTasks.ExecWhenSessionConnectedTask" ThreadingModel="both"/>
      </InProcessServer>
    </Extension>
    <Extension Category="windows.activatableClass.inProcessServer">
      <InProcessServer>
        <Path>SyncFolder\WinRT.Host.dll</Path>
        <ActivatableClass ActivatableClassId="SyncFolder.BackgroundTasks.PortableDeviceWatcherTask" ThreadingModel="both"/>
      </InProcessServer>
    </Extension>
    <Extension Category="windows.activatableClass.inProcessServer">
      <InProcessServer>
        <Path>SyncFolder\WinRT.Host.dll</Path>
        <ActivatableClass ActivatableClassId="SyncFolder.BackgroundTasks.UpdateTask" ThreadingModel="both"/>
      </InProcessServer>
    </Extension>
  </Extensions>

```
Event viewer on the user's computer gives the following error (in the BackgroundTaskInfrastructure section) for any of the 4 tasks when it tries to trigger them (next example is for the 'deviceUse' task, and translated from Dutch):

`An instance of the background task with the access point SyncFolder.BackgroundTasks.PortableDeviceWatcherTask executed for the user ML\mlgoos in session 422859 was returned with error code 0x80008083.`

Did also a test with enabling diagnostics in BackgroundTaskInfrastructure. These are the events being logged when a USB flashdrive is inserted (again translated from Dutch):

```
An activation of a background task with access point SyncFolder.BackgroundTasks.PortableDeviceWatcherTask and name WatchPortableDevices has been placed in the buffer because 2560 has been taken out of the buffer.

An activation of a background task with access point SyncFolder.BackgroundTasks.PortableDeviceWatcherTask and name WatchPortableDevices has been placed in the buffer because 2560 has been taken out of the buffer.

An instance of a background task with access point SyncFolder.BackgroundTasks.PortableDeviceWatcherTask and name WatchPortableDevices was created in session 422859. This was given the id {adac4dc8-f853-40f1-8b45-f562b364fd71}.

An instance of the background task with the access point SyncFolder.BackgroundTasks.PortableDeviceWatcherTask executed for the user ML\mlgoos in session 422859 was returned with error code 0x80008083.
```

The fault occurs on Windows 10, build 19045.3930.

I looked around what 0x80008083 could mean but with no luck. Could it have something to do with OOP BG triggering and the WinRT.Host.dll that is part of the package? Wrong version?

I also checked on another Windows 10 machine with the same build but don't see any issue there. Must however say that on that machine I installed VS Remote tools for debugging. So there might be some packages installed that prevent that the error occurs.

The error occurs on the user's machine when installing from the Microsoft Store but also when sideloading the .msix package.

 Any idea?

### Steps to reproduce the bug

Register for participating in the beta test of the SyncFolder app and install SyncFolder from the Microsoft Store.

### NuGet package version

Windows App SDK 1.4.4: 1.4.231219000

### Packaging type

Packaged (MSIX)

### Windows version

Windows 10 version 22H2 (19045, 2022 Update)

### IDE

Visual Studio 2022

**Comment by DarranRowe** [comment]
I found a reference to this error code in the dotnet/runtime github repository. It was in a file named [host-error-codes.md](https://github.com/dotnet/runtime/blob/main/docs/design/features/host-error-codes.md). That doesn't mean that it is this CoreHostLibMissingFailure error though.
I think your best course of action is to see if you could get a crash dump from the customer.

**Comment by JosHuybrighs** [comment]
@DarranRowe 
Thanks for the prompt answer.

I activated crash dump on the computer but no dump is being generated. I had 'taskbackgroundHost.exe' as key in the registry.

Then started to look deeper in the Windows 10 computer and don't find any of the mentioned dotnet dlls (coreclr, hostfxr, hostpolicy) in there.
Is this normal on a non-developer client computer? On my computers I see these files in various folders and so I am wondering whether this is because I have VS and/or Remote tools installed there.
I can for instance execute 'dotnet' in command prompt on my machines but not on the user's machine. It comes back with 'dotnet' not understood.

Shouldn't a MSIX package install everything that is needed and/or complain when there are some core dlls missing. The strange thing is that the app itself (.NET6) runs without problems. Only OOP background tasks don't run.

**Comment by DarranRowe** [comment]
>Is this normal on a non-developer client computer?
Yes. The dotnet executable is part of the .NET (.NET core) runtime. This doesn't come pre-installed with any of the release versions of Windows. Visual Studio installs this and the .NET SDK when you install any of the .NET workloads. So it is not surprising for a developer system to have dotnet.exe but a non developer system to not have it.

>Shouldn't a MSIX package install everything that is needed and/or complain when there are some core dlls missing.
This is a very much a no.
When Windows installs a MSIX package, the dependencies it checks for are those listed in the package manifest file. If a dependency isn't listed then Windows will not know that the package depends on something. It is possible to experiment with this easily. There are two easy examples of this. The first is to write a C++ application that uses C++ features and then package this up. But make sure that you do not include the dependency on the C++ desktop runtime. Windows will not complain about a lack of dependencies, but the package will not run on a clean install of Windows. The second is .NET.
The issue here is that .NET works in an interesting way. If you look at the package manifest for applications that depend on .NET 5 or newer then you should see that there is no listed dependency for a .NET runtime. As far as I'm aware, there is no framework package for the newer .NET versions either. This means that the runtime must be available in some other way. The options generally are:
1) Have a system installation of .NET.
2) Have the files available locally in the package.
3) Compile everything to native code to remove the dependency.

![Screenshot 2024-01-25 141455](https://github.com/microsoft/WindowsAppSDK/assets/52577874/3e03af55-bc50-4026-babc-0c693d12ca96)

The WinUI 3 projects choose option 2, the .NET runtime is placed in the MSIX package. With how the packaging works, Windows should never end up with an incomplete package install either. There should be an integrity check for the files on install. So the interesting situation here is that if you take any .NET 5 or newer application (even without WinUI 3/Windows App SDK), package this up but do not publish it as self contained then the application will require a system installation of .NET to run.

I suppose one other means of testing this situation is to have a machine with a clean Windows 10 install on it. Make sure to never install .NET or Visual Studio. Then sideload your package and try it out. Hopefully this should place you in the same situation. If it does fail in that situation, another test is to then install the .NET runtime that matches the application's .NET version on the system. If it then works, this would show that the background task engine in Windows isn't using the packaged version of .NET.
Reproducing this issue in a controlled environment is the first step, unfortunately.

>I activated crash dump on the computer but no dump is being generated. I had 'taskbackgroundHost.exe' as key in the registry.

IIRC, .NET uses its own unhandled exception handler and this ignores the registry. Have you tried using the instructions [here](https://learn.microsoft.com/en-us/dotnet/core/diagnostics/collect-dumps-crash)?
